File: src/main/java/com/lcaohoanq/nocket/domain/user/UserPort.java
package com.lcaohoanq.nocket.domain.user;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonInclude.Include;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import com.lcaohoanq.nocket.enums.Gender;
import com.lcaohoanq.nocket.enums.UserRole;
import com.lcaohoanq.nocket.enums.UserStatus;
import jakarta.validation.constraints.Pattern;
import java.time.LocalDateTime;
import java.util.UUID;

public interface UserPort {

    @JsonInclude(Include.NON_NULL)
    record UpdateUserDTO(
        @JsonProperty("name") String name,
        @JsonProperty("email") String email,
        @JsonProperty("phone_number") String phoneNumber,
        @JsonProperty("password") String password,
        @JsonProperty("confirm_password") String confirmPassword, // in case user wants to change password
        @JsonProperty("address") String address,
        @JsonProperty("status")
        @Pattern(regexp = "ACTIVE|INACTIVE|VERIFIED|UNVERIFIED|BANNED", message = "Status must be either ONGOING, INACTIVE, VERIFIED, UNVERIFIED, BANNED")
        String status,
        @JsonProperty("date_of_birth") String dob,
        @JsonProperty("avatar") String avatar
    ) {}

    @JsonPropertyOrder(
        {
            "id",
            "email",
            "password",
            "name",
            "gender",
            "is_active",
            "status",
            "date_of_birth",
            "phone_number",
            "address",
            "avatar",
            "role",
            "wallet_id",
            "preferred_language",
            "preferred_currency",
            "last_login_timestamp",
            "created_at",
            "updated_at"
        }
    )
    record UserResponse(
        UUID id,
        String email,
        @JsonIgnore String password,
        String name,
        Gender gender,
        @JsonProperty("is_active") boolean isActive,
        UserStatus status,
        @JsonProperty("date_of_birth") String dateOfBirth,
        @JsonProperty("phone_number") String phoneNumber,
        @JsonProperty("avatar") String avatar,
        @JsonProperty("role") UserRole role,
        @JsonIgnore @JsonProperty("wallet_id") UUID walletId,
        @JsonProperty("preferred_language") String preferredLanguage,
        @JsonProperty("preferred_currency") String preferredCurrency,

        @JsonProperty("last_login_timestamp")
        @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss.SSSSSS")
        LocalDateTime lastLoginTimestamp,
        @JsonProperty("created_at")
        @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss.SSSSSS")
        LocalDateTime createdAt,
        @JsonProperty("updated_at")
        @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss.SSSSSS")
        LocalDateTime updatedAt
    ) {

    }
    
}

File: src/main/kotlin/com/lcaohoanq/nocket/domain/user/UserPort.kt
package com.lcaohoanq.nocket.domain.user

import com.fasterxml.jackson.annotation.*
import com.lcaohoanq.nocket.enums.Gender
import com.lcaohoanq.nocket.enums.UserRole
import com.lcaohoanq.nocket.enums.UserStatus
import jakarta.validation.constraints.Pattern
import java.time.LocalDateTime
import java.util.*

interface UserPort {
    @JsonInclude(JsonInclude.Include.NON_NULL)
    data class UpdateUserDTO(
        @JsonProperty("name") val name: String,
        @JsonProperty("email") val email: String,
        @JsonProperty("phone_number") val phoneNumber: String,
        @JsonProperty("password") val password: String,
        @JsonProperty("confirm_password") val confirmPassword: String,  // in case user wants to change password
        @JsonProperty("address") val address: String,
        @JsonProperty("status")
        
        @Pattern(
            regexp = "ACTIVE|INACTIVE|VERIFIED|UNVERIFIED|BANNED",
            message = "Status must be either ONGOING, INACTIVE, VERIFIED, UNVERIFIED, BANNED"
        ) val status: String,
        
        @JsonProperty("date_of_birth") val dob: String,
        @JsonProperty("avatar") val avatar: String
    )

    @JsonPropertyOrder(
        "id",
        "email",
        "password",
        "name",
        "gender",
        "is_active",
        "status",
        "date_of_birth",
        "phone_number",
        "address",
        "avatar",
        "role",
        "wallet_id",
        "preferred_language",
        "preferred_currency",
        "last_login_timestamp",
        "created_at",
        "updated_at"
    )
    data class UserResponse(
        val id: UUID,
        val email: String,
        @JsonIgnore val password: String,
        val name: String,
        val gender: Gender,
        @JsonProperty("is_active") val isActive: Boolean,
        val status: UserStatus,
        @JsonProperty("date_of_birth") val dateOfBirth: String,
        @JsonProperty("phone_number") val phoneNumber: String,
        @JsonProperty("avatar") val avatar: String,
        @JsonProperty("role") val role: UserRole,
        @JsonProperty("wallet_id") @JsonIgnore val walletId: UUID,
        @JsonProperty("preferred_language") val preferredLanguage: String,
        @JsonProperty("preferred_currency") val preferredCurrency: String,
        @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss.SSSSSS") @JsonProperty("last_login_timestamp") val lastLoginTimestamp: LocalDateTime,
        @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss.SSSSSS") @JsonProperty("created_at") val createdAt: LocalDateTime,
        @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss.SSSSSS") @JsonProperty("updated_at") val updatedAt: LocalDateTime
    )
}
